### ISSUE TITLE
Incorrect CS8602 warning

### ISSUE THREAD

**Issue by nagya** [dotnet/roslyn#44805]
**Version Used**: 

.NET Core 3.1.300, C# 8.0; issue is a regression since 3.1.201

**Steps to Reproduce**:

```csharp
#nullable enable

namespace ClassLibrary6
{
    public class Class1
    {
        public int A()
        {
            var o = new object(); 
            return C();

            int B()
            {
                while (true) 
                    return o.GetHashCode();
            }
            int C() => B();
        }
    }
}
```

**Expected Behavior**:

No warnings.

**Actual Behavior**:

Class1.cs(15, 28): [CS8602] Dereference of a possibly null reference.

**Notes**:
The bug is quite specific to this code structure. Any of the following changes makes the warning disappear:
- using 'object' instead of 'var'
- omitting the while loop
- moving C before B

**Comment by RikkiGibson** [comment]
I think the problem here is that we analyze nullability in B() and give a diagnostic before observing that C() calls B() and updating the StartingState for B appropriately. At the point where we first analyze B, it basically has an empty StartingState which gets populated with the "default" state of all the variables. 

I think we might be able to improve the situation by "falling back" to a StartingState based on the current State at the point the local function is declared, instead of populating the StartingState with a default state based on variable types.

**Comment by jcouv** [comment]
Interestingly the spurious warning disappears if the two local functions are swapped... ([sharplab](https://sharplab.io/#v2:CYLg1APgxAdgrgGwQQwEYIKYAIMzZgWAChi8BbDAZwAdkBjbAYRUsoBkBLVAJ2W4E8AbMQDexLBKwABAMzSATFmbJWARnGSxRSTulyOMAC5YAggAoAlBt1YtNmwDc+WAPZYAvFhgYA7q9QAVhh0hpYA3FjW9hJSAOxK4cRR0QbGjJYeAHxYAEKJ2tGSqbmWyfZ2hTY+ABYcmFhmhtxwGBaRBZXRca4AdADiGIYAEirVjC7AGPmdWAC+ZRLzHXPEs0A==))

**Comment by RikkiGibson** [comment]
Related to #40157 but not quite the same issue. Here we fail to identify the call to `B()` because it's only called within a local function, not within the call to `C()`.

It's also not clear if using the "state at the point of declaration" for `B()` would cure the issue here. Indeed, it may have a counterproductive effect of suppressing desirable warnings because any local function directly following a return is going to be in an "unreachable" state where everything is assumed to be not-null.

```cs
#nullable enable

namespace ClassLibrary6
{
    public class Class1
    {
        public int A()
        {
            object? o = null;
            return C();

            int B()
            {
                while (true) 
                    return o.GetHashCode(); // no warning
            }
            int C() => B();
        }
    }
}
```

It's not obvious how we can update the starting state of `C()` appropriately before analyzing `C()`, or at least while ensuring that we avoid repeated analysis as much as possible. It's especially flummoxing in cases of mutual recursion.

```cs
#nullable enable

namespace ClassLibrary6
{
    public class Class1
    {
        public int A()
        {
            var o = new object(); 
            return C();

            int B()
            {
                if (GetRandomNumber() == 42) { return C(); }
                while (true) 
                    return o.GetHashCode();
            }
            int C() => B();
        }
    }
}
```

**Comment by zacuke** [comment]
Hello, I am running into this false positive on CS8602

Here is some code sample where it is happening. 

I'm using Visual Studio 17.9.4.
Blazor Web app <Project Sdk="Microsoft.NET.Sdk.Web">
dotnet --list-sdks
6.0.419 [C:\Program Files\dotnet\sdk]
8.0.202 [C:\Program Files\dotnet\sdk]

You can see in picture I have ArgumentNullException.ThrowIfNull(context); - which does resolve CS8602 directly under it, but then in lines further down the warning comes back.

Thanks


![image](https://github.com/dotnet/roslyn/assets/2529117/16b57026-29dc-4124-8526-9b74c355f4ad)

Same code in screenshot but here for text purpose:
```

public async Task InvokeAsync(HttpContext context)
  {
      var httpClient = _httpClientFactory.CreateClient(Strings.ProxyHttpClient);
      var portalKey = context.Request.PathBase.ToString().Split('/')[2];
      var userId = context?.User.FindFirst(ClaimTypes.NameIdentifier)?.Value;
      ArgumentNullException.ThrowIfNull(userId);
      var thisPortal = _db.GetPortal(portalKey, userId) ?? throw new Exception("Portal key not found");
      var remoteHost = thisPortal.RemoteHost.TrimEnd('/');
      var localpath = thisPortal.LocalPath.TrimEnd('/');
      var remotePort = new Uri(remoteHost).Port.ToString();
      ArgumentNullException.ThrowIfNull(context);

      var targetUri = new Uri(remoteHost + context.Request.Path + context.Request.QueryString);

      //var targetRequestMessage = CreateTargetRequestMessage(context.Request, targetUri, remoteHost, remotePort);
      var targetRequestMessage = context?.Request.CreateTargetRequestMessage(targetUri, remoteHost, remotePort, (cookie, newCookies) =>
      {
          if (cookie.Name.StartsWith(".AspNetCore.") )
          {
              //we don't want these getting sent
          }
          else
          {
              newCookies.Add(cookie);
          }
      });
      ArgumentNullException.ThrowIfNull(targetRequestMessage);
      if (targetRequestMessage.Method == HttpMethod.Post)
      {
          targetRequestMessage.Content?.Headers.TryAddWithoutValidation("Content-Type", context.Request.ContentType);
      }
      using var responseMessage = await httpClient.SendAsync(targetRequestMessage, HttpCompletionOption.ResponseHeadersRead, context.RequestAborted);
      context.Response.StatusCode = (int)responseMessage.StatusCode;
      responseMessage.Headers.CopyResponseHeaders(context.Response.Headers, remoteHost, localpath); 

```

**Comment by RikkiGibson** [comment]
context?.Request indicates to the compiler that you think context might be null. Delete the ? and you should start getting the behavior you want.

**Comment by zacuke** [comment]
Oh I see what you're saying, if I remove every last instance of ? the warning goes away. I guess that wasn't apparent, but thank you. I'm trying to force myself to leave <Nullable>enable</Nullable> set for this project, usually I turn it off globally.

I'm trying to remember what originally cause me to go down the rabbit hole of adding ? everywhere. I think I copied some code from somewhere else and then I got a bunch of new warnings below it without realizing.

**Comment by jjonescz** [comment]
Looks like this is fixed by https://github.com/dotnet/roslyn/pull/71679.
